test/framework-ondata-mocha.js

```diff
@@ -0,0 +1,526 @@
+
+               // Verify we got the output state we expect
+               done.vnid.should.equal( '' ); 
+               done.data.should.equal( executedFRunUpdater ); 
+
+               // Cleanup 
+               test.detachAllInputSockets( node );
+            }, function( fail ) { 
+               assert.isNotOk( fail );
+            });
+        });
+
+
+        it( "should fail gracefully when fRunUpdater throws an exception.", function() {
+
+            var inData = "A bit of input data";
+            var executedFRunUpdater = "fRunUpdater failed!!!";
+
+            // Define the fRunUpdater that framework should invoke
+            var fRunUpdater = function() { 
+                throw new Error( executedFRunUpdater );
+            }
+
+            // Create a noflo component and get the node instance for it
+            var wrapper = { fRunUpdater: fRunUpdater };
+            var node = test.createComponent( 
+                factory({
+                          description: "Test Description",
+                          inPorts: {
+                              input: {
+                                  datatype: 'string',
+                                  description: "a string port",
+                                  required: true
+                              }
+                          }
+                        }, 
+                        wrapper )
+            );
+
+            var logBuffer = '';
+            // Hide expected console error message from test output
+            sinon.stub( console, 'error', function (message) {
+                logBuffer += message;
+            }); 
+
+            // Send data to the input port and verify that the fRunUpdater function is called.
+            // We use a promise here because this section is asynchronous
+            return new Promise( function(done, fail) { 
+    
+                test.onOutPortData(node, 'output', done);
+                test.onOutPortData(node, 'error', fail);
+                test.sendData(node, 'input', inData);
+
+            }).then( function( done ) { 
+               // Should NOT succeed here since fRunUpdater threw an exception
+               console.error.restore();
+               assert.isNotOk( done );
+
+            }, function( fail ) { 
+               // Expect a clean failure here since fRunUpdater threw an exception
+               console.error.restore();
+               fail.should.not.be.empty;
+               fail.data.toString().should.equal('Error: '+executedFRunUpdater);
+            });
+        });
+
+        it( "should catch exception when fRunUpdater post processing fails.", function() {
+
+            var inData = "A bit of input data";
+            var executedFRunUpdater = "fRunUpdater success!!!";
+    
+            // Define the fRunUpdater that framework should invoke
+            var fRunUpdater = function() { 
+                // update state
+                this.outputState( undefined );
```

Comment:
This may not always produce an Error at the same place it the code. What about changing the outputState function to throw an error when it is called from within the framework-ondata?